### PR TITLE
#2009 added get user schedule settings endpoint

### LINKED ISSUE #2009: Design Review Calendar - Get User Schedule Settings Endpoint
Blocked by: #1846 

### Description

We need a way to access the schedule settings for a user in Finishline. This ticket is to create the backend endpoint to access the schedule settings a user.

### Acceptance Criteria

- [x] Add an endpoint to retrieve schedule settings
- [x] It should be on the route '/users/:userId/schedule-settings/'
- [x] There should be no payload
- [x] Add tests
- [x] Manually test using Postman, and include screenshots of query and result (reach out to Sean if you have questions)

### Proposed Solution

Take a look at the similar endpoint for retrieving secure settings

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by martin0he**
## Changes

added route to users that returns the specified user's schedule settings

## Test Cases

user has no schedule settings

## Screenshots

Successful Case:
image

User doesn't have settings
image

Submitter isn't the user
image

## To Do

- [x] add condition to return error if user doesn't exist
- [x] add seed data for users with settings for testing
- [x] do postman tests

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #2009

### CODE REVIEW

File: src/backend/tests/users.test.ts [hunk 1]
```diff
@@ -136,4 +138,23 @@ describe('Users', () => {
...
+  describe('getUserScheduleSettings', () => {
+    test('getUserScheduleSettings for user with no settings', async () => {
+      vi.spyOn(prisma.user, 'findUnique').mockResolvedValue(batman);
+      vi.spyOn(prisma.schedule_Settings, 'findUnique').mockResolvedValue(null);
+      await expect(() => UsersService.getUserScheduleSetting(batman.userId)).rejects.toThrow(
+        new HttpException(404, 'User Schedule Settings Not Found')
+      );
+    });
+
+    test('getUserScheduleSettings runs', async () => {
```
Comment (walker-sean) [inline src/backend/tests/users.test.ts:151]: Nitpick: "works" is clearer over "runs"

**Review by walker-sean — CHANGES_REQUESTED** [review]
Nice !

File: src/shared/src/types/user-types.ts [hunk 2]
```diff
@@ -65,3 +65,11 @@ export interface UpdateUserRolePayload {
   userId: number;
   role: string;
 }
+
+export interface userScheduleSettings {
+  drScheduleSettingsId: string;
+  personalGmail: string;
+  personalZoomLink: string;
+  availability: number[];
+  userId: number;
```
[the remark this answers is not in the document]
Comment (martin0he) [reply src/shared/src/types/user-types.ts]: i was copying the same format as secure settings and it includes userId, also I use the userId to find the settings in services

File: src/backend/src/controllers/users.controllers.ts [hunk 3]
```diff
@@ -145,4 +145,14 @@ export default class UsersController {
       next(error);
     }
   }
+
+  static async getUserScheduleSettings(req: Request, res: Response, next: NextFunction) {
+    try {
+      const userId: number = parseInt(req.params.userId);
+      const userScheduleSettings = await UsersService.getUserScheduleSetting(userId);
```
Comment (RChandler234) [inline src/backend/src/controllers/users.controllers.ts:152]: we should pass in the submitter and validate that it's the same as the id of the user who's schedule settings we're requesting. We should probably have a separate endpoint for retrieving availability given a set of userIds

File: src/backend/src/transformers/user-schedule-settings.transformer.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { Prisma } from '@prisma/client';
+import { UserScheduleSettings } from 'shared';
+
+const userScheduleSettingsTransformer = (settings: Prisma.Schedule_SettingsGetPayload<null>): UserScheduleSettings => {
+  return {
+    drScheduleSettingsId: settings.drScheduleSettingsId,
+    personalGmail: settings.personalGmail,
+    personalZoomLink: settings.personalZoomLink,
+    availability: settings.availability,
+    userId: settings.userId
```
Comment (RChandler234) [inline src/backend/src/transformers/user-schedule-settings.transformer.ts:10]: don't include userId here

File: src/shared/src/types/user-types.ts [hunk 5]
```diff
@@ -65,3 +65,11 @@ export interface UpdateUserRolePayload {
   userId: number;
   role: string;
 }
+
+export interface UserScheduleSettings {
+  drScheduleSettingsId: string;
+  personalGmail: string;
+  personalZoomLink: string;
+  availability: number[];
+  userId: number;
```
Comment (RChandler234) [inline src/shared/src/types/user-types.ts:74]: or here

**Review by RChandler234 — CHANGES_REQUESTED** [review]

**Review by RChandler234 — CHANGES_REQUESTED** [review]
The test is failing because the service method returns a shared type but you're comparing it to a prisma type

**Comment by martin0he** [comment]
> The test is failing because the service method returns a shared type but you're comparing it to a prisma type

So should I change the service method to a prisma type or the other one to the shared type (shared is the one without userId and user)?

**Comment by RChandler234** [comment]
> > The test is failing because the service method returns a shared type but you're comparing it to a prisma type
> 
> 
> 
> So should I change the service method to a prisma type or the other one to the shared type (shared is the one without userId and user)?

Change the test to compare the return from the service method to a shared type user schedule settings

File: src/backend/src/routes/users.routes.ts [hunk 6]
```diff
@@ -34,4 +34,5 @@ userRouter.post(
 );
 userRouter.get('/:userId/secure-settings', UsersController.getUserSecureSettings);
```
Comment (walker-sean) [inline src/backend/src/routes/users.routes.ts:36]: remove this newline

File: src/backend/src/services/users.services.ts [hunk 7]
```diff
@@ -320,4 +322,20 @@ export default class UsersService {
...
     return newUserSecureSettings.userSecureSettingsId;
   }
+
+  /**
+   * Gets a user's schedule settings
+   * @param userId the id of the user who's schedule settings are being returned
+   * @returns the user's schedule settings
+   * @throws if the user doesn't have schedule settings
+   */
+  static async getUserScheduleSetting(userId: number, submitter: PrismaUser): Promise<UserScheduleSettings> {
```
Comment (walker-sean) [inline src/backend/src/services/users.services.ts:332]: getUserScheduleSettings (plural)

**Review by walker-sean — CHANGES_REQUESTED** [review]
Looking good!

File: src/backend/src/services/users.services.ts [hunk 8]
```diff
@@ -320,4 +322,20 @@ export default class UsersService {
 
     return newUserSecureSettings.userSecureSettingsId;
   }
+
+  /**
+   * Gets a user's schedule settings
+   * @param userId the id of the user who's schedule settings are being returned
+   * @returns the user's schedule settings
```
Comment (RChandler234) [inline src/backend/src/services/users.services.ts:329]: you should have the submitter param in the comment as well

**Review by RChandler234 — CHANGES_REQUESTED** [review]

**Review by walker-sean — APPROVED** [review]

**Pull request merged**